### ISSUE TITLE
Graphpinator

### ISSUE THREAD

**Issue by peldax** [mwildehahn/hack-graphql#70]
Hello!

I just discovered you made a fork of my project and made some really significant progress into porting, adjusting and improving the functionality for hacklang. Thats really awesome!

Make sure to check out the addons, for example the printer or the constraint directives. Probably not for the code, because you are currently far away from the original PHP source, but maybe some of the ideas may become useful.

I am writing this issue because I am really happy my project made some use for you, and I would like to ask you if you would report possible bugs, inconsistencies and maybe even ideas which you find while working with the codebase. You are possibly the first developers who handled it apart from my colleagues and your feedback would be of a great value for me. I am currently slowly gathering courage to tag 1.0 release and I want everything to fit in perfectly before I finally do it.

Feel free to close the issue.

Thanks!

VP

**Comment by mwildehahn** [comment]
Hey @peldax!

First, just wanted to thank you for writing such a clean set of libraries. We've been trying to leverage GraphQL in our codebase for awhile but the biggest lift has been that we needed one to work with hacklang. The way you broke up Graphpinator into smaller libs made it much easier to port the bits that we needed to get started. So thank you!

We'll report back any issues we discover, so far everything that we've been using from the parser has been ✨ .

**Comment by peldax** [comment]
@mwildehahn Thank you for your kind response!

I am planning for even more subpackages, but its not always easy to completely decouple the dependencies, there seem to always be that one SMALL problem which require some BIG refactoring. 

One great example is introspection and I noticed in one of your issues that you are currently designing it -> And ideally without having the meta fields hardcoded _somewhere_. I am curious about this too, because I did not manage to find a way. It would be really nice to completely move introspection logic & types into separate package, because its otherwise completely standalone, but I did not find a satisfying (not complete overkill) way to decouple the meta fields out from query root type.

Tokenizer and Parser were pretty straight forward because they are low level components, but higher up, when semantic analysis comes into play, its starts to cycle out.

Anyway, I am glad everything was working alright up to this point 👍🏻 .
